File: core/src/main/java/com/cangwang/core/ModuleBus.java
package com.cangwang.core;

import android.content.Context;
import android.content.Intent;
import android.os.Bundle;
import android.support.v4.util.ArrayMap;
import android.util.Log;

import com.cangwang.core.info.MethodInfo;

import java.lang.reflect.Method;
import java.util.ArrayList;
import java.util.Arrays;

/**
 * Created by zjl on 16/10/19.
 */
public class ModuleBus {
    private static final String TAG = "ModuleBus";

    public static final int MODULE_RESULT = 1001;

//    private static ArrayMap<Object,ArrayMap<String,Method>> moduleEventMethods = new ArrayMap<>();
    /**
     * Object methodClass
     * String methodName；
     * MethodInfo method info
     */
    private static ArrayMap<Object,ArrayMap<String,MethodInfo>> moduleEventMethods = new ArrayMap<>();
//    private static ArrayMap<Class<?>,ArrayList<Object>> moduleClients = new ArrayMap<>();

    /**
     * Class IBaseClient.class
     * String methodName
     * Object methodClass
     */
    private static ArrayMap<Class<?>,ArrayMap<String,ArrayList<Object>>> moduleMethodClient = new ArrayMap<>();

    private static ModuleBus instance;

    private String PK_NAME ="";

    public static ModuleBus getInstance(){
        if(instance == null){
            synchronized (ModuleBus.class){
                if (instance == null){
                    instance = new ModuleBus();
                }
            }
        }
        return instance;
    }


    public void register(Object client){
        if(client == null) return;

        Class<?> orginalClass = client.getClass();
        if(orginalClass == null) return;


        Method[] methods = orginalClass.getMethods();

        for(Method method:methods){
            ModuleEvent event = method.getAnnotation(ModuleEvent.class);
            if(event !=null){
                Class<?> clientClass = event.coreClientClass();

//                addClient(clientClass,client);
//                addEventMethod(client,clientClass,method);
                addClient(clientClass, client, method);
//                addEventMethod(clientClass, method);
                addEventMethod(client,method,event.single());
            }
        }
    }

//    private void addClient(Class<?> clientClass,Object client){
//        ArrayList<Object> clientList = moduleClients.get(clientClass);
//
//        if (clientList == null)
//            clientList = new ArrayList<>();
//
//        clientList.add(client);
//        moduleClients.put(clientClass,clientList);
//    }

    private void addClient(Class<?> clientClass,Object client,Method m){
        ArrayMap<String,ArrayList<Object>> clientMethodList = moduleMethodClient.get(clientClass);

        if (clientMethodList == null) {
            clientMethodList = new ArrayMap<>();
        }

        ArrayList<Object> clientList = clientMethodList.get(m.getName());
        if (clientList ==null){
            clientList = new ArrayList<>();
        }

        clientList.add(client);
        clientMethodList.put(m.getName(),clientList);
        moduleMethodClient.put(clientClass,clientMethodList);
    }

//    private void addEventMethod(Object client,Class<?> clientClass, Method m){
//        ArrayMap<String,Method> methods = moduleEventMethods.get(clientClass);
//
//        if(methods == null){
//            methods = new ArrayMap<>();
//            moduleEventMethods.put(clientClass,methods);
//        }
//        methods.put(m.getName(),m);
//    }

//    private void addEventMethod(Class<?> clientClass, Method m){
//        ArrayMap<String,Method> methods = moduleEventMethods.get(clientClass);
//
//        if(methods == null){
//            methods = new ArrayMap<>();
//            moduleEventMethods.put(clientClass,methods);
//        }
//        methods.put(m.getName(),m);
//    }

    private void addEventMethod(Object client, Method m,boolean single){
            ArrayMap<String, MethodInfo> methods = moduleEventMethods.get(client);
            if (methods == null) {
                methods = new ArrayMap<>();
                moduleEventMethods.put(client, methods);
            }
            methods.put(m.getName(), new MethodInfo(m.getName(), m, single));
    }

    public void unregister(Object client){
        if(client == null) return;

        Class<?> orginalClass = client.getClass();
        if(orginalClass == null) return;


        Method[] methods = orginalClass.getMethods();
        for(Method method:methods) {
            ModuleEvent event = method.getAnnotation(ModuleEvent.class);
            if(event !=null) {
                Class<?> clientClass = event.coreClientClass();
                if (moduleEventMethods.get(clientClass)==null) return;
                moduleEventMethods.get(clientClass).remove(method);  //移除方法
//                moduleClients.remove(clientClass);
                if (moduleEventMethods.get(clientClass).isEmpty())   //如果此类中已经无方法，移除此类
                    moduleMethodClient.remove(clientClass);
            }
        }
    }

//    public ArrayList<Object> getClient(Class<?> clientClass){
//
//        if(clientClass == null) return null;
//        ArrayList<Object> clientList = moduleClients.get(clientClass);
//        if(clientList != null){
//            clientList = new ArrayList<>(clientList);
//        }
//
//        return clientList;
//    }

    public ArrayList<Object> getClient(Class<?> clientClass,String methodName){
        if(clientClass == null || methodName == null) return null;
        if (moduleMethodClient.get(clientClass)!=null)
            return moduleMethodClient.get(clientClass).get(methodName);
        else
            return null;
    }

    public void post(Class<?> clientClass,String methodName,Object...args){
        if(clientClass == null || methodName == null ||methodName.length() == 0) return;

        ArrayList<Object> clientList = getClient(clientClass,methodName);

        if(clientList == null) return;

        try{
            for(Object c: clientList){
                try{
                    ArrayMap<String,MethodInfo> methods = moduleEventMethods.get(c);
                    Method method = methods.get(methodName).m;
                    if(method == null){
                        Log.e(TAG,"cannot find client method"+methodName +"for args["+args.length+"]" + Arrays.toString(args));
                        return;
                    }else if(method.getParameterTypes() == null){
                        Log.e(TAG,"cannot find client method param:"+method.getParameterTypes() +"for args["+args.length+"]" + Arrays.toString(args));
                        return;
                    }else if(method.getParameterTypes().length != args.length){
                        Log.e(TAG,"method "+methodName +" param number not matched:method("+method.getParameterTypes().length+"), args(" + args.length+")");
                        return;
                    }
                    method.invoke(c,args);
                }catch (Throwable e){
                    Log.e(TAG,"Notifiy client method invoke error.",e);
                }
            }

        }catch (Throwable e){
            Log.e(TAG,"Notify client error",e);
        }
    }

    public Object postSingle(Class<?> clientClass,String methodName,Object...args){
        if(clientClass == null || methodName == null ||methodName.length() == 0) return null;

        ArrayList<Object> clientList = getClient(clientClass,methodName);

        if(clientList == null) return null;

        try{
            MethodInfo methodInfo = moduleEventMethods.get(clientClass).get(methodName);
            boolean single = methodInfo.single;
            Method method = methodInfo.m;

            if (!single || clientList.size() != 1){
                Log.e(TAG,"method"+methodName +"for args["+args.length+"]" + Arrays.toString(args) + " i s not single");
                return null;
            } else if(method == null){
                Log.e(TAG,"cannot find client method"+methodName +"for args["+args.length+"]" + Arrays.toString(args));
                return null;
            }else if(method.getParameterTypes() == null){
                Log.e(TAG,"cannot find client method param:"+method.getParameterTypes() +"for args["+args.length+"]" + Arrays.toString(args));
                return null;
            }else if(method.getParameterTypes().length != args.length){
                Log.e(TAG,"method "+methodName +" param number not matched:method("+method.getParameterTypes().length+"), args(" + args.length+")");
                return null;
            }

            for(Object c: clientList){
                try{
                    return method.invoke(c,args);
                }catch (Throwable e){
                    Log.e(TAG,"Notifiy client method invoke error.",e);
                }
            }

        }catch (Throwable e){
            Log.e(TAG,"Notify client error",e);
            return null;
        }
        return null;
    }

    private ArrayMap<String,Object> moduleAct = new ArrayMap<>();

    public void startModuleActivity(Object object,String className){
        startModuleActivity(object,className,null);
    }

    public void startModuleActivity(Object object,String className, Bundle bundle){
        if(className == null) return;
        moduleAct.put(className,object);
        post(IBaseClient.class,"startModuleActivity",className,bundle);
    }


    public void moduleResult(Object object,Intent data){
        if (data==null)return;
        post(IBaseClient.class,"moduleResult",moduleAct.get(object.getClass().getName()),data);
    }

    public void setPackageName(String name){
        this.PK_NAME = name;
    }

    public String getPacketName(){
        return PK_NAME;
    }

    public static void init(Context context){
        ModuleCenter.init(context);
    }

}

File: core/src/main/java/com/cangwang/core/ModuleBus.kt
package com.cangwang.core

import android.os.Bundle
import android.content.Context
import android.content.Intent
import android.util.Log
import androidx.collection.ArrayMap
import com.cangwang.core.info.MethodInfo
import org.json.JSONObject
import java.lang.reflect.Method
import java.util.*
import kotlin.reflect.KClass

/**
 * Created by zjl on 16/10/19.
 */
class ModuleBus {
    var packetName = ""
        private set

    fun register(client: Any?) {
        if (client == null) return
        val orginalClass: Class<Any> = client.javaClass ?: return
        val methods = orginalClass.methods
        for (method in methods) {
            val event = method.getAnnotation(ModuleEvent::class.java)
            if (event != null) {
                val clientClass: KClass<*> = event.coreClientClass

//                addClient(clientClass,client);
//                addEventMethod(client,clientClass,method);
                addClient(clientClass, client, method)
                //                addEventMethod(clientClass, method);
                addEventMethod(client, method, event.single)
            }
        }
    }

    //    private void addClient(Class<?> clientClass,Object client){
    //        ArrayList<Object> clientList = moduleClients.get(clientClass);
    //
    //        if (clientList == null)
    //            clientList = new ArrayList<>();
    //
    //        clientList.add(client);
    //        moduleClients.put(clientClass,clientList);
    //    }
    private fun addClient(clientClass: KClass<*>, client: Any, m: Method) {
        var clientMethodList = moduleMethodClient[clientClass]
        if (clientMethodList == null) {
            clientMethodList = ArrayMap()
        }
        var clientList = clientMethodList[m.name]
        if (clientList == null) {
            clientList = ArrayList()
        }
        clientList.add(client)
        clientMethodList[m.name] = clientList
        moduleMethodClient[clientClass] = clientMethodList
    }

    //    private void addEventMethod(Object client,Class<?> clientClass, Method m){
    //        ArrayMap<String,Method> methods = moduleEventMethods.get(clientClass);
    //
    //        if(methods == null){
    //            methods = new ArrayMap<>();
    //            moduleEventMethods.put(clientClass,methods);
    //        }
    //        methods.put(m.getName(),m);
    //    }
    //    private void addEventMethod(Class<?> clientClass, Method m){
    //        ArrayMap<String,Method> methods = moduleEventMethods.get(clientClass);
    //
    //        if(methods == null){
    //            methods = new ArrayMap<>();
    //            moduleEventMethods.put(clientClass,methods);
    //        }
    //        methods.put(m.getName(),m);
    //    }
    private fun addEventMethod(client: Any, m: Method, single: Boolean) {
        var methods = moduleEventMethods[client]
        if (methods == null) {
            methods = ArrayMap()
            moduleEventMethods[client] = methods
        }
        methods[m.name] = MethodInfo(m.name, m, single)
    }

    fun unregister(client: Any?) {
        if (client == null) return
        val orginalClass: Class<Any> = client.javaClass ?: return
        val methods = orginalClass.methods
        for (method in methods) {
            val event = method.getAnnotation(ModuleEvent::class.java)
            if (event != null) {
                val clientClass: KClass<*> = event.coreClientClass
                if (moduleEventMethods[clientClass] == null) return
                moduleEventMethods[clientClass]!!.remove(method.toString()) //移除方法
                //                moduleClients.remove(clientClass);
                if (moduleEventMethods[clientClass]!!.isEmpty) //如果此类中已经无方法，移除此类
                    moduleMethodClient.remove(clientClass)
            }
        }
    }

    //    public ArrayList<Object> getClient(Class<?> clientClass){
    //
    //        if(clientClass == null) return null;
    //        ArrayList<Object> clientList = moduleClients.get(clientClass);
    //        if(clientList != null){
    //            clientList = new ArrayList<>(clientList);
    //        }
    //
    //        return clientList;
    //    }
    fun getClient(clientClass: KClass<*>?, methodName: String?): ArrayList<Any?>? {
        if (clientClass == null || methodName == null) return null
        return if (moduleMethodClient[clientClass] != null) moduleMethodClient[clientClass]!![methodName] else null
    }

    fun post(clientClass: KClass<*>?, methodName: String?, vararg args: Any?) {
        if (clientClass == null || methodName == null || methodName.isEmpty()) return
        val clientList = getClient(clientClass, methodName) ?: return
        try {
            for (c in clientList) {
                val any = try {
                    val methods = moduleEventMethods[c]
                    val method = methods!![methodName]!!.m
                    if (method.parameterTypes.size != args.size) {
                        Log.e(TAG, "method " + methodName + " param number not matched:method(" + method.parameterTypes.size + "), args(" + args.size + ")")
                        return
                    }
                    method.invoke(c, *args)
                } catch (e: Throwable) {
                    Log.e(TAG, "Notifiy client method invoke error.", e)
                }
            }
        } catch (e: Throwable) {
            Log.e(TAG, "Notify client error", e)
        }
    }

    fun postSingle(clientClass: KClass<*>?, methodName: String?, vararg args: Any?): Any? {
        if (clientClass == null || methodName == null || methodName.isEmpty()) return null
        val clientList = getClient(clientClass, methodName) ?: return null
        try {
            val methodInfo = moduleEventMethods[clientClass]!![methodName]
            val single = methodInfo!!.single
            val method = methodInfo.m
            if (!single || clientList.size != 1) {
                Log.e(TAG, "method" + methodName + "for args[" + args.size + "]" + Arrays.toString(args) + " i s not single")
                return null
            } else if (method.parameterTypes.size != args.size) {
                Log.e(TAG, "method " + methodName + " param number not matched:method(" + method.parameterTypes.size + "), args(" + args.size + ")")
                return null
            }
            for (c in clientList) {
                try {
                    return method.invoke(c, *args)
                } catch (e: Throwable) {
                    Log.e(TAG, "Notifiy client method invoke error.", e)
                }
            }
        } catch (e: Throwable) {
            Log.e(TAG, "Notify client error", e)
            return null
        }
        return null
    }

    private val moduleAct = ArrayMap<String, Any>()
    @JvmOverloads
    fun startModuleActivity(`object`: Any, className: String?, bundle: Bundle? = null) {
        if (className == null) return
        moduleAct[className] = `object`
        post(IBaseClient::class, "startModuleActivity", className, bundle)
    }

    fun moduleResult(`object`: Any, data: Intent?) {
        if (data == null) return
        post(IBaseClient::class, "moduleResult", moduleAct[`object`.javaClass.name], data)
    }

    fun setPackageName(name: String) {
        packetName = name
    }

    fun getModuleList(template: String?): List<String?>? {
        return ModuleCenter.getModuleList(template)
    }

    companion object {
        private const val TAG = "ModuleBus"
        const val MODULE_RESULT = 1001
        //    private static ArrayMap<Object,ArrayMap<String,Method>> moduleEventMethods = new ArrayMap<>();
        /**
         * Object methodClass
         * String methodName；
         * MethodInfo method info
         */
        private val moduleEventMethods = ArrayMap<Any, ArrayMap<String?, MethodInfo?>>()
        //    private static ArrayMap<Class<?>,ArrayList<Object>> moduleClients = new ArrayMap<>();
        /**
         * Class IBaseClient.class
         * String methodName
         * Object methodClass
         */
        private val moduleMethodClient = ArrayMap<KClass<*>, ArrayMap<String?, ArrayList<Any?>?>>()
        var instance: ModuleBus? = null
            get() {
                if (field == null) {
                    synchronized(ModuleBus::class.java) {
                        if (field == null) {
                            field = ModuleBus()
                        }
                    }
                }
                return field
            }
            private set

        fun init(context: Context) {
            ModuleCenter.init(context)
        }

        fun init(context: Context?, `object`: JSONObject?) {
            ModuleCenter.init(context, `object`, true)
        }
    }
}